src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,305 @@
+
+                let gio_file = gio::File::for_path(&document_local.path);
+
+                imp.file_button.connect_clicked(move |_|{
+                    gio::AppInfo::launch_default_for_uri(&gio_file.uri(), Option::<&gio::AppLaunchContext>::None).ok();
+                });
+
+            } else if !document_local.is_downloading_active {
+                imp.file_status
+                    .set_icon_name(Some("document-save-symbolic"));
+
+                let id = data.document.document.id;
+
+                let session = Rc::new(message.chat().session());
+
+                imp.file_button.connect_clicked(
+                    clone!(@weak self as this, @strong session => move |_| {
+                        this.download_document(id, &session);
+                    }),
+                );
+            }
+        }
+    }
+
+    fn download_document(&self, file_id: i32, session: &Session) {
+        println!("downloading");
+
+        let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
+
+        receiver.attach(
+            None,
+            clone!(@weak self as obj => @default-return glib::Continue(false), move |file| {
+                let imp = obj.imp();
+
+                if file.local.is_downloading_completed {
+                    imp.file_status.set_icon_name(Some("folder-documents-symbolic"));
+                    imp.file_row.set_subtitle(human_readable_size(file.local.downloaded_size).as_str());
+
+                    println!("Downloaded file path: {}", &file.local.path);
+                    
+                    let gio_file = gio::File::for_path(file.local.path);
+
+                    imp.file_button.connect_clicked(move |_|{
+                        gio::AppInfo::launch_default_for_uri(&gio_file.uri(), Option::<&gio::AppLaunchContext>::None).ok();
+                    });
+                } else {
+                    let progress = file.local.downloaded_size as f64 / file.expected_size as f64;
+                    
+                    imp.file_row.set_subtitle(format!("Downloading: {}% of {}", progress * 100.0, human_readable_size(file.expected_size)).as_str());
+                }
+
+                glib::Continue(true)
+            }),
+        );
+
+        session.download_file(file_id, sender);
+    }
+}
+
+fn shorten_string(name: String) -> String {
+    match name.chars().count() {
+        0..=47 => name,
+        len => {
+            let start = name.chars().take(22);
+            let end = name.chars().skip(len - 22);
+            String::from_iter(start.chain("...".chars()).chain(end))
+        }
+    }
+}
+
+fn human_readable_size(file_size: i32) -> String {
+    let suffix = ["B", "KB", "MB", "GB"]; // Need to localize this
+    let index = match file_size.leading_zeros() {
+        22..=32 => 0,
+        12..=21 => 1,
+        2..=11 => 2,
+        _ => 3,
+    };
+
+    format!("{} {} ", file_size >> (10 * index), suffix[index])
+}
```

Comment:
There is also `glib::format_size` :-)

Reply:
Thank you, I thought the suffix needed to be translated, but I found that in telegram clients, it is always in English.
I will use glib::format_size

Reply:
> There is also `glib::format_size` :-)

TIL :)